### ISSUE TITLE
Replace message priority with multiple ports for transmission

### ISSUE THREAD

**Issue by Qiao-Jin** [neo-project/neo#1874]
**Summary or problem description**

Currently we use high/low priority to distinguish between different kinds of messages. We insert messages into 2 different queues, high/low priority queue, and send messages according to following rule:

(1) Depending upon message type, push the message into corresponding queue (high / low priority)

(2) If there are no messages in either queue or we have not received acknowledge signal of last message to the neighboring node, return

https://github.com/neo-project/neo/blob/7032907cbbebd9189bcd01e6b83328b864a0a5b2/src/neo/Network/P2P/RemoteNode.cs#L49

(3) Otherwise, if there are messages in high priority queue, send the earliest message from high priority queue. If not , send the earliest message from low priority queue

**Such rule has many limitations.** Transmission efficiency can be very low, as before sending any new messages, **we need to wait until receiving the acknowledge signal of last message**. Furthermore, even though there are 2 priority queue for different messages, in fact all messages are transmitted through 1 physical channel. The result is that, **even a high priority message (i.e. Consensus Payload) might be queued in message receiver's mailbox for a long time before handled.** 

**Do you have any solution you want to propose?**

Instead of keeping 2 priority queue, we can use 2 connections for high / low priority messages, respectively. For each connection, we don't need to keep local message queues. Instead we shall send messages directly, and still be sure that high priority messages are handled in the first place.

In Akka-remote, different `ActorSystem `are bound to different ports respectively. In Akka-remote's config we can define the binding address/port of `ActorSystem`. Furthermore, it's also enabled to send messages to remote actors directly, which will make transmission logic much simpler. We can also send messages in batch in Akka-remote and get better transmission capability as described in #1718. So to replace priority queue, we can define 2 `ActorSystem`s,  `LowPrioritySystem` (which serves transaction and corresponding `inv` and `getdata` messages) `HighPrioritySystem` (which serves other messages), and construct different actors with different `ActorSystem`s. We can send messages directly to remote actors in batches rather than one-by-one through a commom tcp manager.

By testing, after replacing Akka with Akka-remote and rewriting current transmission logic, what is achieved is as follows:

(1) Much better transmission capability. Transactions sent/received per second is much higher.

(2) High priority messages will no longer be delayed. Consensus payload will always be on time so no ChangeView even for heavy transaction traffic scenarios. Other messages such as `getdata `message will also be received and handled in time so that block sychronization is much smoother even for heavy transaction traffic scenarios. (Currently, in case of heavy transaction scenarios, new created block will almost never reach other nodes. The reason is that, the mailbox of consensus node's remotenode actor is full of transaction messages so that the `getdata` messages from other nodes will be queued far behind in the mailbox.)

**Neo Version**
- Neo 3

**Where in the software does this update applies to?**
- Network Policy
- P2P (TCP)

**Comment by roman-khimov** [comment]
> we need to wait until receiving the acknowledge signal of last message

Can anyone really explain why sends are being blocked by ACKs (which are TCP thing actually we should not care about at all)? This logic was introduced in #288, but I'm unable to find any discussions around this specific topic. I only see it as a thing limiting the number of in-flight messages and thus trying to prevent high-priority message delays. So it's trading bandwidth for a bit more predictable delays. But the cost of doing so in terms of bandwidth is huge and I don't think it's worth it. Have you tried just removing this logic?

**Comment by erikzhang** [comment]
Akka-remote uses other protocol instead of neo protocol to deliver messages.

**Comment by Qiao-Jin** [comment]
>  Have you tried just removing this logic?

If we only remove this logic, it will mean that priority queue will be meaningless, as every message will be sent immediately. The result is that, high priority messages will not arrive in time in case of heavy transaction traffic, which will give rise to:

(1) Consensus node will be always trying to changing view instead of reach any consensus

(2) Ping/pong messages not able to reach in time, which means synchronization will fail

... etc.

**Comment by shargon** [comment]
> every message will be sent immediately

This should be perfect :P

I think that we can remove the `ack` condition, but not the `verack`

**Comment by Qiao-Jin** [comment]
> > every message will be sent immediately
> 
> This should be perfect :P
> 
> I think that we can remove the `ack` condition, but not the `verack`

We cannot just remove the `ack` condition... unless we replace priority queue with mutiple `ActorSystem` at the same time

**Comment by erikzhang** [comment]
> every message will be sent immediately

The messages won't be sent immediately. They will be enqueued into the tcp actor without priority.

**Comment by Qiao-Jin** [comment]
> > every message will be sent immediately
> 
> The messages won't be sent immediately. They will be enqueued into the tcp actor without priority.

We can use different actors which is constructed with 2 different `ActorSystem`s so that priority is still preserved.

**Comment by erikzhang** [comment]
Maybe we can configure custom mailboxes for tcp actors.

https://getakka.net/articles/actors/mailboxes.html

**Comment by Qiao-Jin** [comment]
> Maybe we can configure custom mailboxes for tcp actors.
> 
> https://getakka.net/articles/actors/mailboxes.html

I think we have already defined kinds of PriorityMailbox such as ConsensusServiceMailbox?

**Comment by erikzhang** [comment]
> I think we have already defined kinds of PriorityMailbox such as ConsensusServiceMailbox?

Yes. But you need to configure the tcp actores to use the mailbox.
